godot-core/src/builtin/quaternion.rs

```diff
@@ -311,6 +342,12 @@ impl std::fmt::Display for Quaternion {
     }
 }
 
+impl From<Basis> for Quaternion {
+    fn from(basis: Basis) -> Self {
+        basis.to_quat()
+    }
+}
```

Comment:
I think this does give the user more control, this provides both Quaternion::from or basis.into() too. Depending on the user's coding style and what's easier for them to understand, any of the above may be better for them.

Reply:
Multiple points here:
- New features need to have a justification to exist, satisfying all code styles isn't a good one. If we followed that principle, we'd have a cluttered mess in no time :wink: 
- Redundant features need an ever higher bar. Once there are multiple ways to achieve the same thing, the whole "which should I use" and "is there a difference" questions come up.
- `From` has the problem that it cannot do type inference. You can write `let a = basis.to_quat()` but not `let a = basis.into()` (at least not without further info after this line).
- When you use `into()`, your code is objectively harder to understand than a named method that shows _what_ you convert into. The main benefit in my opinion is for generic programming, which isn't applicable here.
- One argument is that `From` demonstrates equivalence between two types. This _also_ doesn't apply here, because basis can store rotation + scale, while quaternion can only store rotation. This even goes against not one, but **three** [official recommendations](https://doc.rust-lang.org/std/convert/trait.From.html#when-to-implement-from) stating that `From` should be lossless, value-preserving and obvious.
- godot-rust as part of its API design doesn't use `From` in many places. Traditionally we've had it for string types, and it helped keep arguments generic as `impl Into<GString>`. Now there's `AsArg` which takes over that role, but some conversions are still done via `From`. Maybe named methods would be better there too, who knows.

This discussion comes up every now and then, I should probably add it to the book :slightly_smiling_face:

Reply:
> * when using into one can easily say into::

Actually not, the way it's defined doesn't allow you to use turbofish `.into::<Quaternion>()`. It's another common pitfall -- because the `into()` function isn't generic itself, only the [`Into<T>`](https://doc.rust-lang.org/std/convert/trait.Into.html) trait is :slightly_smiling_face: 



> to_quat is a specific function to know, it could be to_quaternion instead, it's not as intuitive imo.

Fair point :thinking: it was added in https://github.com/godot-rust/gdext/pull/124 alongside `Basis::from_quat` for the other direction. In the context it's probably quite clear, there are also other abbreviations in conversions, such as `from_cols`, `into_dyn` etc. But I'm also OK with changing it.


> Thank you for illustrating me on this and sorry for having said something that dumb.

It's not dumb at all! I appreciate any input as long as it's constructive, and the fact that this has been brought up a few time shows that it's a frequent question, too. So no worries at all :blush: 

> * I think it would be a nice addition to the book? A conversions between types section, maybe.

This particular instance is probably best kept in the API docs (as this is the place for looking up API of a concrete type). But what can make sense is to elaborate some general "conventions" or API designs that godot-rust uses. I'll make some notes about this.

Reply:
> It's another common pitfall
Oh goddess Hylia I think I hate into now. I would then just advocate for a To<Type> trait with to<Type>() function, that makes the "to type" action a standard action in the repository, but I'm not against it not being a trait, I just think it could be easier to find the one function you need from a user standpoint.

> But I'm also OK with changing it.
I'm not saying from_quat and to_quat are bad options though, I'm all on board of using abbreviations when it's clear. I just wish for times when there would be function aliases since what one people thinks obvious another would not. I'm not a library maintainer so I don't know much about these though.

> So no worries at all 😊
I'm glad I don't annoy with my comments at the least.

> But what can make sense is to elaborate some general "conventions" or API designs that godot-rust uses. I'll make some notes about this.
Yeah! This would be great, specially for contributors. I don't mind much so long as there's a convention I can follow safely and know users will understand.